File: system/ulib/cobalt-client/counter.cpp
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <string.h>

#include <cobalt-client/cpp/counter-internal.h>
#include <cobalt-client/cpp/counter.h>
#include <zircon/assert.h>

namespace cobalt_client {
namespace internal {

RemoteCounter::RemoteCounter(const RemoteMetricInfo& metric_info)
    : BaseCounter(), metric_info_(metric_info) {
    buffer_ = 0;
}

RemoteCounter::RemoteCounter(RemoteCounter&& other)
    : BaseCounter(fbl::move(other)), buffer_(fbl::move(other.buffer_)),
      metric_info_(other.metric_info_) {}

bool RemoteCounter::Flush(Logger* logger) {
    // Write the current value of the counter to the buffer, and reset it to 0.
    buffer_ = this->Exchange();
    return logger->Log(metric_info_, buffer_);
}

void RemoteCounter::UndoFlush() {
    this->Increment(buffer_);
}

} // namespace internal

void Counter::Increment(Counter::Count value) {
    remote_counter_.Increment(value);
}

Counter::Count Counter::GetRemoteCount() const {
    return remote_counter_.Load();
}

} // namespace cobalt_client

File: src/sys/lib/stress-test/src/counter.rs
// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    fuchsia_async::Task,
    futures::{channel::mpsc, StreamExt},
    log::debug,
    std::collections::HashMap,
};

/// An unbounded mpsc channel connecting each runner to the counter thread.
///
/// CounterTx is cloned and given to all runners. Upon a successful operation, the runner
/// will send their actor's name over the channel.
///
/// The counter thread receives the actors name and updates the current operation count.
/// This ensures that runners are not blocked.
pub type CounterTx = mpsc::UnboundedSender<String>;

/// Starts a new task that maintains a count of all successful operations
/// This task will terminate when the target number of operations has been hit.
///
/// Returns the counter task and a CounterTx.
pub fn start_counter(target: u64) -> (Task<()>, CounterTx) {
    let (tx, mut rx) = mpsc::unbounded();
    let task = Task::spawn(async move {
        // Keep track of global count + individual actor contributions
        let mut count_map: HashMap<String, u64> = HashMap::new();
        let mut total = 0;

        // Run this task until the count has been met
        while total < target {
            // Wait for an actor to finish an operation
            let key = rx.next().await.unwrap();

            // Update the actor's contribution
            if let Some(value) = count_map.get_mut(&key) {
                *value += 1;
            } else {
                count_map.insert(key, 1);
            }

            // Update global count
            total += 1;

            debug!("Counters -> [total:{}] {:?}", total, count_map);
        }
    });
    (task, tx)
}
